### ISSUE TITLE
Removing old buffer support

### ISSUE THREAD

**Issue by methane** [python/cpython#85275]
BPO | [41103](https://bugs.python.org/issue41103)
--- | :---
Nosy | @vstinner, @encukou, @methane, @ambv, @hroncok, @miss-islington
PRs | <li>python/cpython#21117</li><li>python/cpython#27437</li><li>python/cpython#27441</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = 'https://github.com/encukou'
closed_at = None
created_at = <Date 2020-06-24.13:43:42.640>
labels = ['expert-C-API', '3.10', '3.11']
title = 'Removing old buffer support'
updated_at = <Date 2021-08-05.08:33:02.859>
user = 'https://github.com/methane'
```

bugs.python.org fields:
```python
activity = <Date 2021-08-05.08:33:02.859>
actor = 'petr.viktorin'
assignee = 'petr.viktorin'
closed = False
closed_date = None
closer = None
components = ['C API']
creation = <Date 2020-06-24.13:43:42.640>
creator = 'methane'
dependencies = []
files = []
hgrepos = []
issue_num = 41103
keywords = ['patch']
message_count = 12.0
messages = ['372251', '372391', '381854', '381865', '381873', '398459', '398460', '398467', '398476', '398477', '398560', '398974']
nosy_count = 7.0
nosy_names = ['vstinner', 'petr.viktorin', 'methane', 'lukasz.langa', 'hroncok', 'miss-islington', 'tarun.johar']
pr_nums = ['21117', '27437', '27441']
priority = 'normal'
resolution = 'wont fix'
stage = 'resolved'
status = 'open'
superseder = None
type = None
url = 'https://bugs.python.org/issue41103'
versions = ['Python 3.10', 'Python 3.11']
```

</p></details>

### Linked PRs
* gh-105137

**Comment by vstinner** [comment]
I closed my issue #105186 as a duplicate of this one. Copy of my message.

---

Would it be ok to remove the legacy "buffer" API in Python 3.13, scheduled for October 2024? I'm talking about functions:

* PyObject_AsCharBuffer()
* PyObject_CheckReadBuffer()
* PyObject_AsReadBuffer()
* PyObject_AsWriteBuffer()

This API is deprecated since Python 3.0  and is older than the new ``Py_buffer`` API. The ``Py_buffer`` API is safer: it has a PyBuffer_Release() method to notify when the consumer is done: the resource can be released. The ``Py_buffer`` API is now commonly used.

In June 2020, there was a first attempt to remove these functions in issue #85275: commit 6f8a6ee59cb7f99f68df8ee9c3e8c8cf19af3eed. It caused too many troubles and had to be reverted the following year (commit ce5e1a6809b714eb0383219190a076d9f883e008, July 2021). The removed functions were still used by at least 6 projects:

* PyQt4: **FIXED** in PyQt5 by using the safe PyObject_GetBuffer()/PyBuffer_Release() functions. PyQt4 is now legacy.
* libsolv: **FIXED**. it seems like it has been solved in https://github.com/openSUSE/libsolv/issues/410
* m2crypto: **FIXED** in m2crypto 0.37.1 by https://gitlab.com/m2crypto/m2crypto/-/commit/6bb65c27f01dd69f08c4654ce531fb283c7bcc07
* apsw: **FIXED** in 3.35.4-r1, https://github.com/rogerbinns/apsw/issues/301
* wsaccel: **FIXED** in v0.6.3: https://github.com/methane/wsaccel/issues/22 and https://github.com/methane/wsaccel/commit/b171314d883df04bf5449085e017e44dee21231f
* djvulibre: It seems like [it stills uses PyObject_AsWriteBuffer()](https://github.com/jwilk-archive/python-djvulibre/blob/181ed40594562296d74cdb5bd91fb7b964eae25b/djvu/common.pxi#L70). But https://github.com/jwilk-archive/python-djvulibre project has been archived by the owner on Oct 4, 2022 and is now read-only!

The webassets project was also mentioned, but I'm not sure that its failure is related to these removal: see https://bugzilla.redhat.com/show_bug.cgi?id=1899555 I cannot find these functions in https://github.com/Kronuz/pyScss code right now.

Sadly, the 4 functions are part of the stable ABI and have to be kept there. I'm only asking to remove them in the C API. The issue #88775 was complaining that the functions were removed wheareas they are part of the stable ABI.

**Comment by encukou** [comment]
Sorry I didn't get to this.
Removing them from public API, including the limited API, is AFAIK fine. But

> Sadly, the 4 functions are part of the stable ABI and have to be kept there

Please keep (or add) tests for them, so they don't regress.
